q2_diversity_lib/tests/test_util.py

```diff
@@ -39,9 +47,137 @@ def test_pass_empty_table_as_kwarg(self):
+    package = 'q2_diversity_lib.tests'
+
+    def setUp(self):
+        super().setUp()
+        self.empty_table = biom.Table(np.array([]), [], [])
+        # empty table generated from self.empty_table with biom v2.1.7
+        self.empty_table_fp = self.get_data_path('empty_table.biom')
+
+        @_disallow_empty_tables_passed_filepath
+        def f1(table: biom.Table):
+            pass
+        self.function_with_table_param = f1
+        self.test_function = f1
+
+        @_disallow_empty_tables_passed_filepath
+        def f2():
+            pass
+        self.function_without_table_param = f2
+
+    def test_pass_empty_table_positionally(self):
+        with self.assertRaisesRegex(ValueError, "table.*is empty"):
+            self.function_with_table_param(self.empty_table_fp)
+
+    def test_pass_empty_table_as_kwarg(self):
+        with self.assertRaisesRegex(ValueError, "table.*is empty"):
+            self.function_with_table_param(table=self.empty_table_fp)
+
+    def test_decorated_lambda_with_table_param(self):
+        with self.assertRaisesRegex(ValueError, "table.*is empty"):
+            decorated_lambda = _disallow_empty_tables_passed_filepath(
+                        lambda table: None)
+            decorated_lambda(self.empty_table_fp)
+
+    def test_wrapped_function_has_no_table_param(self):
+        with self.assertRaisesRegex(TypeError, "no parameter.*table"):
+            self.function_without_table_param()
+
+    def test_passed_object_not_filepath(self):
+        with self.assertRaisesRegex(TypeError, "path should be string"):
+            self.test_function(table=self.empty_table)
+
+
+class SafelyConstrainNJobsTests(TestPluginBase):
+    package = 'q2_diversity_lib.tests'
+
+    def setUp(self):
+        super().setUp()
+
+        @_safely_constrain_n_jobs
+        def function_no_params():
+            pass
+        self.function_no_params = function_no_params
+
+        @_safely_constrain_n_jobs
+        def function_w_param(n_jobs=3):
+            return n_jobs
+        self.function_w_n_jobs_param = function_w_param
+
+        self.valid_table_fp = self.get_data_path('two_feature_table.biom')
+        self.valid_tree_fp = self.get_data_path('three_feature.tree')
+        self.valid_table = biom.load_table(self.valid_table_fp)
+
+    def test_function_without_n_jobs_param(self):
+        with self.assertRaisesRegex(TypeError, 'without \'n_jobs'):
+            self.function_no_params()
+
+    @mock.patch("q2_diversity_lib._util.psutil.Process")
+    def test_function_with_an_n_jobs_param(self, mock_cpu_affinity):
+        mock_cpu_affinity = psutil.Process()
+        mock_cpu_affinity.cpu_affinity = mock.MagicMock(return_value=[0, 1, 2])
+        self.assertEqual(self.function_w_n_jobs_param(3), 3)
+
+    @mock.patch('psutil.Process.cpu_affinity', side_effect=AttributeError)
+    @mock.patch('psutil.cpu_count', return_value=999)
+    def test_system_has_no_cpu_affinity(self, mock_cpu_count, mock_cpu_affin):
```

Comment:
This test fails on macos, I suspect something is funky with the monkeypatch.

Reply:
I took a closer look at the docs ([link](https://docs.python.org/3/library/unittest.mock.html#patch)), and noticed this:

> By default patch() will fail to replace attributes that don’t exist. If you pass in create=True, and the attribute doesn’t exist, patch will create the attribute for you when the patched function is called, and delete it again after the patched function has exited. This is useful for writing tests against attributes that your production code creates at runtime. It is off by default because it can be dangerous. With it switched on you can write passing tests against APIs that don’t actually exist!

Tossing a `create=True` into the `patch` appears to make this test pass. I double checked that the side effect you specified was actually being propagated through, by replacing your `ValueError` with other errors, and when I do the test fails (as expected), because the trap you've built in `_safely_constrain_n_jobs` fails to catch  whatever new error symbol I tossed in. Jackpot!

For example, using this code:

```patch
diff --git a/q2_diversity_lib/tests/test_util.py b/q2_diversity_lib/tests/test_util.py
index dfa9c1c..8bd6fc8 100644
--- a/q2_diversity_lib/tests/test_util.py
+++ b/q2_diversity_lib/tests/test_util.py
@@ -106,7 +106,7 @@ class SafelyConstrainNJobsTests(TestPluginBase):
         self.assertEqual(self.function_w_n_jobs_param(3), 3)
 
     @mock.patch("q2_diversity_lib._util.psutil.Process.cpu_affinity",
-                side_effect=AttributeError)
+                side_effect=IndexError, create=True)
     @mock.patch('psutil.cpu_count', return_value=999)
     def test_system_has_no_cpu_affinity(self, mock_cpu_count, mock_cpu_affin):
         self.assertEqual(self.function_w_n_jobs_param(999), 999)
```

yields the follow (expected) error:

```bash
$ py.test -k test_system_has_no_cpu_affinity
============================================================================================================================================= test session starts ==============================================================================================================================================
platform darwin -- Python 3.6.10, pytest-5.4.2, py-1.8.1, pluggy-0.13.1
rootdir: /Users/matthew/projects/qiime2/q2-diversity-lib
collected 46 items / 45 deselected / 1 selected                                                                                                                                                                                                                                                                

q2_diversity_lib/tests/test_util.py F                                                                                                                                                                                                                                                                    [100%]

=================================================================================================================================================== FAILURES ===================================================================================================================================================
__________________________________________________________________________________________________________________________ SafelyConstrainNJobsTests.test_system_has_no_cpu_affinity ___________________________________________________________________________________________________________________________

self = <q2_diversity_lib.tests.test_util.SafelyConstrainNJobsTests testMethod=test_system_has_no_cpu_affinity>, mock_cpu_count = <MagicMock name='cpu_count' id='4747156728'>, mock_cpu_affin = <MagicMock name='cpu_affinity' id='4823751984'>

    @mock.patch("q2_diversity_lib._util.psutil.Process.cpu_affinity",
                side_effect=IndexError, create=True)
    @mock.patch('psutil.cpu_count', return_value=999)
    def test_system_has_no_cpu_affinity(self, mock_cpu_count, mock_cpu_affin):
>       self.assertEqual(self.function_w_n_jobs_param(999), 999)

q2_diversity_lib/tests/test_util.py:112: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
<decorator-gen-526>:2: in function_w_param
    ???
q2_diversity_lib/_util.py:70: in _safely_constrain_n_jobs
    cpus = len(psutil.Process().cpu_affinity())
../../../.conda/envs/q2dev/lib/python3.6/unittest/mock.py:939: in __call__
    return _mock_self._mock_call(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

_mock_self = <MagicMock name='cpu_affinity' id='4823751984'>, args = (), kwargs = {}, self = <MagicMock name='cpu_affinity' id='4823751984'>, _call = call(), seen = set(), do_method_calls = False, method_call_name = 'cpu_affinity', mock_call_name = '', is_a_call = False, _new_parent = None

    def _mock_call(_mock_self, *args, **kwargs):
        self = _mock_self
        self.called = True
        self.call_count += 1
    
        # handle call_args
        _call = _Call((args, kwargs), two=True)
        self.call_args = _call
        self.call_args_list.append(_call)
    
        seen = set()
    
        # initial stuff for method_calls:
        do_method_calls = self._mock_parent is not None
        method_call_name = self._mock_name
    
        # initial stuff for mock_calls:
        mock_call_name = self._mock_new_name
        is_a_call = mock_call_name == '()'
        self.mock_calls.append(_Call(('', args, kwargs)))
    
        # follow up the chain of mocks:
        _new_parent = self._mock_new_parent
        while _new_parent is not None:
    
            # handle method_calls:
            if do_method_calls:
                _new_parent.method_calls.append(_Call((method_call_name, args, kwargs)))
                do_method_calls = _new_parent._mock_parent is not None
                if do_method_calls:
                    method_call_name = _new_parent._mock_name + '.' + method_call_name
    
            # handle mock_calls:
            this_mock_call = _Call((mock_call_name, args, kwargs))
            _new_parent.mock_calls.append(this_mock_call)
    
            if _new_parent._mock_new_name:
                if is_a_call:
                    dot = ''
                else:
                    dot = '.'
                is_a_call = _new_parent._mock_new_name == '()'
                mock_call_name = _new_parent._mock_new_name + dot + mock_call_name
    
            # follow the parental chain:
            _new_parent = _new_parent._mock_new_parent
    
            # check we're not in an infinite loop:
            # ( use ids here so as not to call __hash__ on the mocks)
            _new_parent_id = id(_new_parent)
            if _new_parent_id in seen:
                break
            seen.add(_new_parent_id)
    
        effect = self.side_effect
        if effect is not None:
            if _is_exception(effect):
>               raise effect
E               IndexError

../../../.conda/envs/q2dev/lib/python3.6/unittest/mock.py:999: IndexError
```

 I haven't wrapped my head around the implications of adding this `create=True` flag, but it is pointing in the right direction.

I wish psutil would just include this `cpu_affinity` attribute on all OSes, and just raise a `NotImplemented` on the hosts that didn't support it. I imagine they have their reasons for implementing it the way they did, though.

Reply:
Oh, I should've mentioned, this was in response to the following action run:

https://github.com/qiime2/q2-diversity-lib/actions/runs/117324052